core/src/main/java/org/elasticsearch/action/admin/indices/shards/TransportIndicesShardStoresAction.java

```diff
@@ -213,7 +213,7 @@ void finish() {
              * A shard exists/existed in a node only if shard state file exists in the node
              */
             private boolean shardExistsInNode(final NodeGatewayStartedShards response) {
-                return response.storeException() != null || response.version() != -1;
+                return response.storeException() != null || response.version() != -1 || response.allocationId() != null;
```

Comment:
I'm not sure `response.allocationId() != null`  is OK. What happens when we allocate old indices that don't have allocationId serialized?